triedb/pathdb/lookup.go

```diff
@@ -0,0 +1,271 @@
+// This function assumes that all layers older than the provided one have already
+// been processed, ensuring that layers are processed strictly in a bottom-to-top
+// order.
+func (l *lookup) addLayer(diff *diffLayer) {
+	defer func(now time.Time) {
+		lookupAddLayerTimer.UpdateSince(now)
+	}(time.Now())
+
+	var (
+		wg    sync.WaitGroup
+		state = diff.rootHash()
+	)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for accountHash := range diff.states.accountData {
+			list, exists := l.accounts[accountHash]
+			if !exists {
+				list = make([]common.Hash, 0, 16) // TODO(rjl493456442) use sync pool
+			}
+			list = append(list, state)
+			l.accounts[accountHash] = list
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for accountHash, slots := range diff.states.storageData {
+			subset := l.storages[accountHash]
+			if subset == nil {
+				subset = make(map[common.Hash][]common.Hash)
+				l.storages[accountHash] = subset
+			}
+			for slotHash := range slots {
+				list, exists := subset[slotHash]
+				if !exists {
+					list = make([]common.Hash, 0, 16) // TODO(rjl493456442) use sync pool
+				}
+				list = append(list, state)
+				subset[slotHash] = list
+			}
+		}
+	}()
+	wg.Wait()
+}
+
+// removeLayer traverses the state data retained in the specified diff layer and
+// unlink them from the lookup set.
+func (l *lookup) removeLayer(diff *diffLayer) error {
+	defer func(now time.Time) {
+		lookupRemoveLayerTimer.UpdateSince(now)
+	}(time.Now())
+
+	var (
+		wg    errgroup.Group
+		state = diff.rootHash()
+	)
+	wg.Go(func() error {
+		for accountHash := range diff.states.accountData {
+			var (
+				found bool
+				list  = l.accounts[accountHash]
+			)
+			// Traverse the list from oldest to newest to quickly locate the ID
+			// of the stale layer.
+			for i := 0; i < len(list); i++ {
+				if list[i] == state {
+					if i == 0 {
+						list = list[1:]
+						if cap(list) > 1024 {
+							list = append(make([]common.Hash, 0, len(list)), list...)
+						}
+					} else {
+						list = append(list[:i], list[i+1:]...)
+					}
+					found = true
+					break
+				}
+			}
+			if !found {
+				return fmt.Errorf("account lookup is not found, %x, state: %x", accountHash, state)
+			}
+			if len(list) != 0 {
+				l.accounts[accountHash] = list
+			} else {
+				delete(l.accounts, accountHash)
+			}
+		}
+		return nil
+	})
+
+	wg.Go(func() error {
+		for accountHash, slots := range diff.states.storageData {
+			subset := l.storages[accountHash]
+			if subset == nil {
+				return fmt.Errorf("storage lookup is not found, %x", accountHash)
+			}
+			for slotHash := range slots {
+				var (
+					found bool
+					list  = subset[slotHash]
+				)
+				// Traverse the list from oldest to newest to quickly locate the ID
+				// of the stale layer.
+				for i := 0; i < len(list); i++ {
+					if list[i] == state {
+						if i == 0 {
+							list = list[1:]
+							if cap(list) > 1024 {
+								list = append(make([]common.Hash, 0, len(list)), list...)
```

Comment:
If we have 1024 non-finalized diff layers floating around, we would hit this every time, right? 
I get the idea of pruning away some unnecessary memory if we ever hit a lot of parallel forks. I guess its not really a big deal either way

Reply:
> If we have 1024 non-finalized diff layers floating around

The number of diff layers is defined as 128, regardless of the associated state is finalized or not. 
But for sure, with this PR, we can unlock the potential to pile up arbitrary number of diff layers 
on top (if the memory is sufficient to host that many diff layers). In such cases, we need to adjust
this constant.

The main motivation to add this additional check is: for some storage slots frequently being updated,
it might be modified in every blocks. The underlying slice might be held forever. This mechanism just
tries to make sure the dangling slices won't be referenced forever.